image-registry/automate-pruning.md

```diff
@@ -0,0 +1,252 @@
+  - "@bparees"
+  - "@coreydaley"
+  - "@ricardomarascini"
+  - "@gabemontero"
+approvers:
+  - "@bparees"
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+creation-date: 2019-09-26
+last-updated: 2019-19-26
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Automate Image Pruning
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This enhancement will install a `CronJob` to automate the image pruning task
+for the internal registry. An API will be exposed to allow cluster admins to
+configure the pruning job to suit their needs, or disable pruning if that is
+what is desired.
+
+## Motivation
+
+Regular image pruning is a vital maintenance task to ensure an OpenShift
+cluster remains healthy. In certain use cases (such as CI/build farms), pruning
+must be run frequently to prevent etcd from exhausting storage. Today, pruning
+is configured through a manual, multi-step process
+
+### Goals
+
+- Provide a one-step process for cluster admins to regularly schedule image
+pruning.
+- Default pruning to be enabled on an OpenShift cluster.
+- Provide alerting if a cluster is at risk due to pruning failing/too many
+images.
+
+### Non-Goals
+
+- Fix fundamental issues with the current pruning process (ex: pruning images
+that may be referenced in a CRD).
+
+## Proposal
+
+### User Stories
+
+#### Prune images in etcd
+
+As an OpenShift cluster admin
+I want the cluster to automatically prune ImageStreams stored in etcd
+So that etcd storage is not overwhelmed by unused imagestreams
+And the cluster remains healthy
+
+#### Prune images on the registry
+
+As an OpenShift cluster admin
+I want the cluster to automatically prune unused images stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+#### Orphaned blobs
+
+As an OpenShift cluster admin
+I want the cluster to automatically remove orphaned blobs stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+### Implementation Details/Notes/Constraints
+
+The implementation relies the current image pruning process invoked by the
+`oc adm prune` command. Running this in a container can be accomplished by
+using the `cli` imagestream installed by the samples operator. The image must
+be run using a service account with the `system:image-pruner` cluster role.
+
+The image registry operator will be updated to install the following:
+
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable parameters:
```

Comment:
I wouldn't expect the operator to directly install the cronjob, i'd expect it to bootstrap a pruning configuration (possibly part of the registry operator resource) on initial install, and then i'd expect the operator to take that pruning configuration and produce/reconcile/update a cronjob to accomplish the requested pruning config.

Reply:
@bparees I don't understand the different between "directly install the cronjob" and "produce/reconcile/update a cronjob", could you elaborate on this?

Reply:
it might just be the way i was parsing the description, but the section says:

`The image registry operator will be updated to install the following:`

Some of the things it lists sound like things i'd expect to be in the manifest for the operator (thus created by the CVO when the operator is installed/created.  e.g. service accounts, rolebindings).

But the CronJob definition would not be in the manifests...  i would expect the cronjob to be created/managed by the operator in the same way the the operator creates/manages the registry deployment.

So i'd expect the flow to be:

1) CVO installs the manifest resources (service accounts, roles/bindings, operator deployment, etc)

2) operator comes up, creates its own registry config resource (e.g. the registry storage configuration today) which would now include configuration for automated pruning  (what we call bootstrapping its config)

3) operator sees its configuration, which would include config for automated pruning, and create a cronjob w/ the appropriate args (and on updates to that configuration, would sync any changes to args to that cronjob).

Reply:
@bparees updated language to make it clear it's the _operator's_ job to create and reconcile, not the CVO:
```
The image registry operator will be updated to configure and reconcile all the
components needed to run image pruning on an OpenShift cluster.
```

Reply:
well, the SA and rolebindings will be reconciled by the CVO assuming you create them as manifest resources.

but i think the current phrasing covers the key part which is how the cronjob is managed.